File: app/composeApp/src/commonMain/kotlin/com/enconomy/pop/chain/PopCtx.kt
package com.enconomy.pop.chain

import com.enconomy.pop.hexToBytes
import com.enconomy.pop.toHex
import org.kotlincrypto.hash.sha3.Keccak256

/**
 * PopCtx (docs/worldid/01 §5): the one context-binding formula.
 *
 *   session_nonce = keccak256(abi.encode(bytes32 DOMAIN, uint256 chainId, address consumer,
 *                                        bytes32 ctxHash, uint64 notBefore, bytes16 sid))
 *   signal        = session_nonce ‖ role (33 B, role 'A' 0x41 | 'B' 0x42)
 *   signalHash    = uint256(keccak256(signal)) >> 8
 *
 * The 192-byte preimage is built by hand as six 32-byte words; no ABI library.
 */
object PopCtx {
    const val PREIMAGE_LEN = 192

    fun keccak(b: ByteArray): ByteArray = Keccak256().digest(b)

    /** keccak256("pop-ctx-v1"), the hash, not the string. */
    val DOMAIN: ByteArray = keccak("pop-ctx-v1".encodeToByteArray())

    fun preimage(chainId: Long, consumer: ByteArray, ctxHash: ByteArray, notBefore: Long, sid: ByteArray): ByteArray {
        require(chainId >= 0) { "chainId < 0" }
        require(consumer.size == 20) { "consumer must be 20 bytes" }
        require(ctxHash.size == 32) { "ctxHash must be 32 bytes" }
        require(notBefore >= 0) { "notBefore < 0" }
        require(sid.size == 16) { "sid must be 16 bytes" }
        val out = ByteArray(PREIMAGE_LEN)
        DOMAIN.copyInto(out, 0)                    // word 0
        putBe(out, 64, chainId)                   // word 1: uint256, left zero pad
        consumer.copyInto(out, 64 + 12)           // word 2: 12 zero bytes ‖ address
        ctxHash.copyInto(out, 96)                 // word 3
        putBe(out, 160, notBefore)                // word 4: uint64, left zero pad
        sid.copyInto(out, 160)                    // word 5: bytes16 left-aligned, right zero pad
        check(out.size == PREIMAGE_LEN)
        return out
    }

    fun nonce(chainId: Long, consumer: ByteArray, ctxHash: ByteArray, notBefore: Long, sid: ByteArray): ByteArray =
        keccak(preimage(chainId, consumer, ctxHash, notBefore, sid))

    /** Wire form: consumer "0x"+40 hex, ctxHash "0x"+64 hex, sessionId 32 hex (no prefix). Returns 32 bytes. */
    fun nonce(chainId: Long, consumer: String, ctxHash: String, notBefore: Long, sessionId: String): ByteArray =
        nonce(chainId, hex0x(consumer, 20), hex0x(ctxHash, 32), notBefore, hexN(sessionId.removePrefix("0x"), 16))

    fun signal(nonce: ByteArray, role: Char): ByteArray {
        require(nonce.size == 32) { "nonce must be 32 bytes" }
        require(role == 'A' || role == 'B') { "role must be A or B" }
        return nonce + byteArrayOf(role.code.toByte())
    }

    /** The IDKit signal string: lowercase "0x" + 66 hex, so IDKit decodes it to raw bytes. */
    fun signalHex(nonce: ByteArray, role: Char): String = "0x" + signal(nonce, role).toHex()

    /** uint256(keccak256(b)) >> 8 as "0x%064x". */
    fun fieldHash(b: ByteArray): String = "0x00" + keccak(b).copyOf(31).toHex()

    fun signalHash(nonce: ByteArray, role: Char): String = fieldHash(signal(nonce, role))

    fun action(sessionId: String): String = "pop:$sessionId"

    fun actionField(action: String): String = fieldHash(action.encodeToByteArray())

    /** Big-endian [v] in the 32-byte word that ends at [end] (exclusive). */
    private fun putBe(out: ByteArray, end: Int, v: Long) {
        for (i in 0 until 8) out[end - 1 - i] = (v ushr (8 * i)).toByte()
    }

    private fun hex0x(s: String, n: Int): ByteArray {
        require(s.startsWith("0x") || s.startsWith("0X")) { "missing 0x" }
        return hexN(s.substring(2), n)
    }

    private fun hexN(s: String, n: Int): ByteArray {
        require(s.length == 2 * n && s.all { it in '0'..'9' || it in 'a'..'f' || it in 'A'..'F' }) { "want $n hex bytes" }
        return s.lowercase().hexToBytes()
    }
}

File: server/pop/popctx.py
"""PopCtx (worldid spec 01 §5): the one context-binding formula.

session_nonce = keccak256(abi.encode(bytes32 DOMAIN, uint256 chainId, address consumer,
                                     bytes32 ctxHash, uint64 notBefore, bytes16 sid))
Preimage is built by hand as 6 x 32-byte words (192 bytes). keccak is pycryptodome's Keccak-256,
not hashlib.sha3_256 (different padding).
"""
from __future__ import annotations

from Crypto.Hash import keccak as _keccak


def keccak(b: bytes) -> bytes:
    h = _keccak.new(digest_bits=256)
    h.update(b)
    return h.digest()


DOMAIN = keccak(b"pop-ctx-v1")
ROLE_BYTE = {"A": b"\x41", "B": b"\x42"}


def _addr(a: str | bytes) -> bytes:
    raw = bytes.fromhex(a[2:] if a[:2] in ("0x", "0X") else a) if isinstance(a, str) else a
    if len(raw) != 20:
        raise ValueError("address must be 20 bytes")
    return raw


def _b32(x: str | bytes) -> bytes:
    raw = bytes.fromhex(x[2:] if x[:2] in ("0x", "0X") else x) if isinstance(x, str) else x
    if len(raw) != 32:
        raise ValueError("ctx_hash must be 32 bytes")
    return raw


def preimage(chain_id: int, consumer: str | bytes, ctx_hash: str | bytes, not_before: int, sid: str | bytes) -> bytes:
    sid_b = bytes.fromhex(sid) if isinstance(sid, str) else sid
    if len(sid_b) != 16:
        raise ValueError("sid must be 16 bytes")
    if not (0 <= not_before < 1 << 64) or not (0 <= chain_id < 1 << 256):
        raise ValueError("chain_id/not_before out of range")
    pre = (DOMAIN
           + chain_id.to_bytes(32, "big")
           + bytes(12) + _addr(consumer)
           + _b32(ctx_hash)
           + not_before.to_bytes(32, "big")        # uint64, left zero pad
           + sid_b + bytes(16))                    # bytes16, left-aligned, right zero pad
    assert len(pre) == 192
    return pre


def nonce(chain_id: int, consumer: str | bytes, ctx_hash: str | bytes, not_before: int, sid: str | bytes) -> bytes:
    return keccak(preimage(chain_id, consumer, ctx_hash, not_before, sid))


def signal(n: bytes, role: str | bytes) -> bytes:
    """abi.encodePacked(bytes32 session_nonce, bytes1 role): 33 bytes."""
    rb = ROLE_BYTE[role] if isinstance(role, str) else role
    assert len(n) == 32 and rb in (b"\x41", b"\x42")
    return n + rb


def signal_str(n: bytes, role: str | bytes) -> str:
    """The IDKit signal string: lowercase 0x + even-length hex (decoded to raw bytes by IDKit)."""
    return "0x" + signal(n, role).hex()


def signal_hash_hex(b: bytes) -> str:
    return "0x%064x" % (int.from_bytes(keccak(b), "big") >> 8)


def action(sid: str) -> str:
    return "pop:" + sid


def action_field_hex(sid: str) -> str:
    return signal_hash_hex(action(sid).encode())
